src/frontend/pages/article/table_of_contents.rs

```diff
@@ -0,0 +1,160 @@
+#[derive(PartialEq)]
+struct Toc {
+    entries: Vec<TocEntry>,
+}
+
+impl Toc {
+    fn count_entries_with_level(&self, level: u32) -> usize {
+        self.entries.iter().filter(|e| e.level == level).count()
+    }
+}
+
+#[derive(PartialEq)]
+struct TocEntry {
+    level: u32,
+    sec_number: String,
+    name: String,
+    id: String,
+    children: Toc,
+}
+
+#[derive(PartialEq)]
+pub(crate) struct TocBuilder {
+    top_level: Toc,
+    chain: Vec<TocEntry>,
+}
+
+impl TocBuilder {
+    pub(crate) fn new() -> TocBuilder {
+        TocBuilder {
+            top_level: Toc {
+                entries: Vec::new(),
+            },
+            chain: Vec::new(),
+        }
+    }
+
+    fn into_toc(mut self) -> Toc {
+        self.fold_until(0);
+        self.top_level
+    }
+
+    fn fold_until(&mut self, level: u32) {
+        let mut this = None;
+        loop {
+            match self.chain.pop() {
+                Some(mut next) => {
+                    next.children.entries.extend(this);
+                    if next.level < level {
+                        self.chain.push(next);
+                        return;
+                    } else {
+                        this = Some(next);
+                    }
+                }
+                None => {
+                    self.top_level.entries.extend(this);
+                    return;
+                }
+            }
+        }
+    }
+
+    pub(crate) fn push(&mut self, level: u32, name: String, id: String) -> &str {
+        assert!(level >= 1);
+
+        self.fold_until(level);
+
+        let mut sec_number;
+        {
+            let toc = match self.chain.last() {
+                None => {
+                    sec_number = String::new();
+                    &self.top_level
+                }
+                Some(entry) => {
+                    sec_number = entry.sec_number.clone();
+                    sec_number.push('.');
+                    &entry.children
+                }
+            };
+            let number = toc.count_entries_with_level(level);
+            sec_number.push_str(&(number + 1).to_string())
+        }
+
+        self.chain.push(TocEntry {
+            level,
+            name,
+            sec_number,
+            id,
+            children: Toc {
+                entries: Vec::new(),
+            },
+        });
+
+        let just_inserted = self.chain.last_mut().unwrap();
+        &just_inserted.sec_number
+    }
+}
+
+impl Toc {
+    fn print_inner(&self, v: &mut String) {
+        use std::fmt::Write as _;
+
+        if !&self.entries.is_empty() {
+            v.push_str("<ul>");
+            for entry in &self.entries {
+                let _ = write!(
+                    v,
+                    "\n<li><a href=\"#{id}\">{num} {name}</a>",
```

Comment:
The links are not working, it ends up generating simply `<a href="#">2.1 3</a>`. The anchor link is usually simply `#{name}`, but in case of duplicate  titles it appends a number. 

https://github.com/markdown-it-rust/markdown-it-plugins.rs/blob/main/crates/heading_anchors/src/lib.rs